### PR TITLE
Feature/scopechange/workflow

### PR DESCRIPTION
**Pull request by Gustav-Eikaas**
# Description

Please include a summary of the change and which issue is fixed.
List any external dependencies that are required for this change.

Completes: AB#FILL_IN_YOUR_ISSUE_ID

## Checklist:

-   [ ] I have performed a self-review of my own code.
-   [ ] I have linked my DevOps task using the AB# tag.
-   [ ] My code is easy to read, and comments are added where needed.
-   [ ] My code is covered by tests.

### CODE REVIEW

File: src/apps/ScopeChangeRequest/Api/ScopeChange/Request/initiateScopeChange.ts [hunk 1]
```diff
@@ -34,4 +34,7 @@ export async function initiateScopeChange(request: ScopeChangeRequest): Promise<
     };
 
     await patchScopeChange(payload);
+
+
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Api/ScopeChange/Request/initiateScopeChange.ts:38]: remove

File: src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestEditForm.tsx [hunk 2]
```diff
@@ -15,9 +24,68 @@ export const ScopeChangeRequestEditForm = ({
...
+                object: x,
+                searchValue: x.procosysNumber,
+                type: 'tag',
+            })
+        );
+
+        request.documents.forEach((x) =>
+            relations.push({
+                label: `DOC_${x.stidDocumentNumber}`,
+                value: x.stidDocumentNumber,
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestEditForm.tsx:66]: function

File: src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestEditForm.tsx [hunk 3]
```diff
@@ -15,9 +24,68 @@ export const ScopeChangeRequestEditForm = ({
...
+                object: x,
+                searchValue: x.stidDocumentNumber,
+                type: 'document',
+            })
+        );
+
+        // request.areas.forEach((x) => relations.push({
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestEditForm.tsx:73]: uncomment

File: src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestEditForm.tsx [hunk 4]
```diff
@@ -27,21 +95,41 @@ export const ScopeChangeRequestEditForm = ({
...
+            documentNumbers: documents.map((x) => x.value) || [],
+            areaCodes: areas.map((x) => x.value) || [],
+            disciplineCodes: disciplines.map((x) => x.value) || [],
+        });
+
+        attachments.forEach(async (attachment) => {
+            await uploadAttachment({ requestId: request.id, file: attachment });
         });
+
+        if (!error) cancel();
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestEditForm.tsx:120]: rename

File: src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestForm.tsx [hunk 5]
```diff
@@ -70,17 +75,21 @@ export const ScopeChangeRequestForm = ({
         );
         if (scID) {
             attachments.forEach(async (attachment) => {
-                await uploadAttachment(scID, attachment);
+                await uploadAttachment({
+                    file: attachment,
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Form/ScopeChangeRequestForm.tsx:79]: useMutation

File: src/apps/ScopeChangeRequest/Components/SearchableDropdown/RelatedObjectsSearch/RelatedObjectsSearch.tsx [hunk 6]
```diff
@@ -118,6 +121,15 @@ export const RelatedObjectsSearch = ({
                     apiErrors.map((name) => {
                         return <ErrorWrapper key={name}>Failed to fetch {name}</ErrorWrapper>;
                     })}
+                <TitleBar>
+                    <Title>References</Title>
+
+                    <StidSelector
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/SearchableDropdown/RelatedObjectsSearch/RelatedObjectsSearch.tsx:127]: Advanced search

File: src/apps/ScopeChangeRequest/Components/Sidesheet/ErrorBanner.tsx [hunk 7]
```diff
@@ -11,25 +14,46 @@ export interface ErrorFormat {
...
+                            {message?.validationErrors && Object.values(message.validationErrors).map((errorArray) => {
+                                return (
+                                    <>
+                                        {errorArray.map((error) => <Banner.Message key={error}>{error}</Banner.Message>)}
+                                    </>
+                                )
+                            })}
+                        </ErrorMessageContainer>
+                        <Banner.Actions>
+                            <Button onClick={() => setErrorMessage(null)}>Dismiss</Button>
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Sidesheet/ErrorBanner.tsx:46]: set message to null

File: src/apps/ScopeChangeRequest/Components/Sidesheet/ScopeChangeSidesheet.tsx [hunk 8]
```diff
@@ -36,54 +37,68 @@ export const ScopeChangeSideSheet = (item: ScopeChangeRequest): JSX.Element => {
             retry: false,
             onError: () =>
                 setErrorMessage({
-                    message: ' Network error, please check your connection and try again',
+                    message: {
+                        detail: "Failed to connect to server",
+                        statusCode: 0,
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Sidesheet/ScopeChangeSidesheet.tsx:42]: Make fields optional

File: src/apps/ScopeChangeRequest/Components/Sidesheet/ScopeChangeSidesheet.tsx [hunk 9]
```diff
@@ -36,54 +37,68 @@ export const ScopeChangeSideSheet = (item: ScopeChangeRequest): JSX.Element => {
...
-    async function notifyChange() {
-        await queryClient.invalidateQueries(QueryKeys.History);
-        await queryClient.invalidateQueries(QueryKeys.Scopechange);
-    }
+    const notifyChange = useCallback(async () => {
+        await queryClient.fetchQuery(QueryKeys.History);
+        await queryClient.fetchQuery(QueryKeys.Scopechange);
+    }, [queryClient]);
 
     const refetchScopeChange = async () => {
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Sidesheet/ScopeChangeSidesheet.tsx:65]: useCallback

File: src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/AddContributor.tsx [hunk 10]
```diff
@@ -27,12 +29,9 @@ export const AddContributor = ({ onCancel }: AddContributorProps): JSX.Element =
...
     };
 
-    const { mutateAsync, isLoading } = useMutation(submit, {
-        onSettled: async () => {
-            await notifyChange();
-            onCancel();
-        },
-        onError: () => setErrorMessage('Failed to assign contributor'),
+    const { mutateAsync, isLoading } = useScopeChangeMutation(submit, {
+        onSuccess: () => onCancel(),
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/AddContributor.tsx:33]: close

File: src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/SignWithComment.tsx [hunk 11]
```diff
@@ -65,7 +65,10 @@ export const SignWithComment = ({ criteria, onCancel }: SignWithCommentProps): J
                 />
             </Section>
             <ButtonContainer>
-                <Button disabled={!text || text.length === 0} onClick={() => mutateAsync()}>
+                <Button
+                    disabled={!text || text.length === 0}
+                    onClick={() => mutateAsync({ action: 'Approved' }).then(() => onCancel())}
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/SignWithComment.tsx:70]: onCancel = close

File: src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+}
+
+export const WorkflowCriteria = ({ step, criteria, canAddContributor }: WorkflowCriteriasProps): JSX.Element => {
+    const [selected, setSelected] = useState<TypedSelectOption | null>(null);
+    const { request, setErrorMessage } = useScopeChangeContext();
+    const [signComment, setSignComment] = useState<string>('');
+    const [showSignWithComment, setShowSignWithComment] = useState(false);
+
+    const checkCanSign = () => canSign({ criteriaId: criteria.id, requestId: request.id, stepId: step.id })
+    const { data: userCanSign } = useQuery(`step/${step.id}/criteria/${criteria.id}`, checkCanSign, { refetchOnWindowFocus: false });
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx:38]: Convert to hook

File: src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+                        ? 'RequireProcosysFunctionalRoleSignature'
+                        : 'RequireProcosysUserSignature'
+                        }`,
+                    value: selected.value,
+                },
+            });
+            setSelected(null);
+            setShowReassign(false);
+        }
+    }, [selected]);
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx:124]: add deps

File: src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+
+    const { mutateAsync: signMutation, isLoading: signLoading, error } = useScopeChangeMutation(
+        onSignStep,
+        {
+            onError: (e: ServerError) => setErrorMessage(e),
+            onSuccess: () => setSignComment("")
+        }
+    );
+
+    useEffect(() => {
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx:186]: remove

File: src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+                    <MenuButton
+                        items={makeSignOptions()}
+                        onMenuOpen={() => closeAll()}
+                        buttonText="Sign"
+                    />
+
+                    <IconMenu items={makeMoreActions()} onMenuOpen={() => closeAll()} />
+                </Inline>
+            </WorkflowStepViewContainer>
+            <Loading />
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Components/Workflow/Criteria/Components/WorkflowCriteria.tsx:222]: inline ternary

File: src/apps/ScopeChangeRequest/Functions/Access/getContributionId.ts [hunk 16]
```diff
@@ -13,7 +13,7 @@ export function getContributionId(
 ): string | undefined {
     const foundContributor = contributors
         .filter((contributor) => contributor.contribution === null)
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/Functions/Access/getContributionId.ts:15]: Delete file

File: src/apps/ScopeChangeRequest/ScopeChangeRequestApp.tsx [hunk 17]
```diff
@@ -49,15 +53,44 @@ export function setup(appApi: ClientApi): void {
...
+                                <div>{cell.row.original.currentWorkflowStep.name}</div>
+                            ) : (
+                                ''
+                            )}
+                        </div>
+                    );
+                },
+                id: 'CurrentStep',
+                width: 180,
+                Aggregated: () => console.log('-'),
```
Comment (Gustav-Eikaas) [inline src/apps/ScopeChangeRequest/ScopeChangeRequestApp.tsx:88]: null

**Review by jaysencpp — APPROVED** [review]

**Pull request merged**